optaplanner-test/src/main/java/org/optaplanner/test/api/score/stream/SingleConstraintAssertion.java

```diff
@@ -39,9 +43,27 @@
     }
 
     private Number getImpact() {
-        return constraintMatchTotalMap.values().stream()
-                .mapToInt(ConstraintMatchTotal::getConstraintMatchCount)
-                .sum();
+        Score<?> zero = getParentConstraintVerifier().getConstraintStreamScoreDirectorFactory()
+                .getScoreDefinition()
+                .getZeroScore();
+        if (constraintMatchTotalMap.isEmpty()) {
+            return zero.toLevelNumbers()[0]; // Return zero in the numeric type expected by the caller.
+        }
+        // We do not know the matchWeight, so we need to deduce it.
+        // Constraint matches give us a score, whose levels are in the form of (matchWeight * constraintWeight).
+        // Here, we strip the constraintWeight.
+        Score<?> totalMatchWeightedScore = constraintMatchTotalMap.values().stream()
+                .map(s -> s.getScore().divide(s.getConstraintWeight()))
+                .reduce(zero, Score::add);
+        // Each level of the resulting score now has to be the same number, the matchWeight.
+        List<Number> matchWeightsFound = Arrays.stream(totalMatchWeightedScore.toLevelNumbers())
+                .distinct()
+                .collect(Collectors.toList());
+        if (matchWeightsFound.size() != 1) {
+            throw new IllegalStateException("Programming error: only one match weight should have been found ("
+                    + matchWeightsFound + ").");
```

Comment:
The score here is `matchWeight * constraintWeight * matchCount`.
At this point, I have already divided by it `constraintWeight`.
What remains is therefore `matchWeight * matchCount`. And therefore the value in every level of the score must be the same.

I believe there already is an explanatory comment in the code.
(That said, you correctly point out a bug in case where one level of the score is `0`. Fixing that.)

Reply:
For example, consider `.penalize("something", HardSoftScore.of(1, 2), something -> 2)`.
Further consider that this penalty is only applied once. In this case:

- `matchCount = 1`
- `matchWeight = 2`
- `constraintWeight = 1hard/2soft`

And therefore, from the end:
- `constraintMatchScore = 2hard/4soft`
- `constraintMatchScore / constraintWeight = 2hard/2soft`
- `matchWeightsFound = [2]`
- `result = 2`

Reply:
Beautiful explanation, thanks! Now it makes perfect sense.